arangod/IResearch/IResearchAqlAnalyzer.h

```diff
@@ -0,0 +1,109 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2020 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Andrei Lobov
+////////////////////////////////////////////////////////////////////////////////
+#ifndef ARANGOD_IRESEARCH__IRESEARCH_CALCULATION_ANALYZER
+#define ARANGOD_IRESEARCH__IRESEARCH_CALCULATION_ANALYZER 1
+
+#include "analysis/analyzers.hpp"
+#include "analysis/token_attributes.hpp"
+#include "Aql/Ast.h"
+#include "Aql/AqlItemBlockManager.h"
+#include "Aql/ExecutionEngine.h"
+#include "Aql/ExecutionPlan.h"
+#include "Aql/QueryContext.h"
+#include "Aql/SharedAqlItemBlockPtr.h"
+#include "StorageEngine/TransactionState.h"
+
+#include <string>
+
+namespace arangodb {
+namespace iresearch {
+
+class AqlAnalyzer final : public irs::analysis::analyzer{
+ public:
+  struct Options {
+    Options() = default;
+
+    Options(std::string&& query, bool collapse, bool keep, uint32_t batch)
+      : queryString(query), collapsePositions(collapse),
+      keepNull(keep), batchSize(batch) {}
+
+    /// @brief Query string to be executed for each document.
+    /// Field value is set with @param binded parameter.
+    std::string queryString;
+
+    /// @brief determines how processed members of array result:
+    /// if set to true all members are considered to be at position 0
+    /// if set to false each array members is set at positions serially
+    bool collapsePositions{ false };
+
+    /// @brief do not emit empty token if query result is NULL
+    /// this could be used fo index filtering.
+    bool keepNull{ true };
+
+    /// @brief  batch size for running query. Set to 1 as most of the cases
+    /// we expect just 1 to 1 modification query.
+    uint32_t batchSize{ 1 };
```

Comment:
this highly differs from the default batchSize in AQL queries (1000 rows).
Many of the Aql Blocks do not define how many rows they need to produce, and will result to ask for 1000 rows upstream, causing the upstream to pre-allocate a 1000 lines and then just use 1.
There are some places in AQL where we can scale this down, but not always.

So in order to make this more efficient I think we need to make the default-batch-size internally adjustable to a more sensible number for this case. (I would not set it to 1 that will be too small, but maybe ~10 is best value for this case)